service/frontend/nexus_incoming_service_registry.go

```
@@ -0,0 +1,406 @@
+	if !r.isTableVersionCurrent(ctx) {
+		r.invalidateCachedServices()
+		if err := r.fetchServices(ctx); err != nil {
+			return nil, fmt.Errorf("error loading nexus incoming services frontend cache: %w", err)
+		}
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	service, ok := r.servicesByName[request.Name]
+	if !ok {
+		return nil, serviceerror.NewNotFound(fmt.Sprintf("frontend could not find cached service with name %v", request.Name))
+	}
+
+	return &operatorservice.GetNexusIncomingServiceResponse{Service: service}, nil
+}
+
+func (r *registry) ListNexusIncomingServices(
+	ctx context.Context,
+	request *operatorservice.ListNexusIncomingServicesRequest,
+) (*operatorservice.ListNexusIncomingServicesResponse, error) {
+	if !r.isTableVersionCurrent(ctx) {
+		r.invalidateCachedServices()
+		if err := r.fetchServices(ctx); err != nil {
+			return nil, fmt.Errorf("error loading nexus incoming services frontend cache: %w", err)
+		}
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	startIdx := 0
+	if request.NextPageToken != nil {
+		nextServiceID := string(request.NextPageToken)
+
+		startFound := false
+		startIdx, startFound = slices.BinarySearch(r.serviceIDsSorted, nextServiceID)
+
+		if !startFound {
+			return nil, serviceerror.NewFailedPrecondition("could not find service indicated by nexus list incoming services next page token")
+		}
+	}
+
+	endIdx := startIdx + int(request.PageSize)
+	if endIdx > len(r.serviceIDsSorted) {
+		endIdx = len(r.serviceIDsSorted)
+	}
+
+	var nextPageToken []byte
+	if endIdx < len(r.services) {
+		nextPageToken = []byte(r.serviceIDsSorted[endIdx])
+	}
+
+	return &operatorservice.ListNexusIncomingServicesResponse{
+		NextPageToken: nextPageToken,
+		Services:      r.services[startIdx:endIdx],
+	}, nil
+}
+
+func (r *registry) isTableVersionCurrent(ctx context.Context) bool {
+	actualTableVersion, err := r.persistence.GetNexusIncomingServicesTableVersion(ctx)
+	if err != nil {
+		return false
```

You should return the error and not assume this is false.